cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -0,0 +1,148 @@
+use std::path::PathBuf;
+
+use thiserror::Error;
+
+use super::build::ManifestBuilder;
+use super::catalog::Client;
+use super::git::GitCommandProvider;
+use crate::flox::FloxhubToken;
+use crate::models::environment::managed_environment::ManagedEnvironment;
+use crate::models::environment::path_environment::PathEnvironment;
+use crate::models::lockfile::Lockfile;
+
+pub enum PublishEnvironment {
+    Path(PathEnvironment),
+    Managed(ManagedEnvironment),
+}
+
+/// The `Publish` trait describes the high level behavior of publishing a package to a catalog.
+/// Authentication, upload, builds etc, are implementation details of the specific provider.
+/// Modeling the behavior as a trait allows us to swap out the provider, e.g. a mock for testing.
+pub trait Publish {
+    fn publish(
+        &self,
+        client: Client,
+        builder: &dyn ManifestBuilder,
```

Comment:
```suggestion
        builder: &dyn ManifestBuilder,
```
I would prefer we not use a trait object here. You can use `impl ManifestBuilder` instead.